src/devices/Card/Ultralight/UltralightCard.cs

```diff
@@ -0,0 +1,995 @@
+                    }
+                }
+            }
+            else
+            {
+                return false;
+            }
+
+            BlockNumber++;
+            Data![0] = 0x03; // NDEF start marker
+            Data[1] = 0x00; // Length of message
+            Data[2] = 0xFE; // NDEF end marker
+            Data[3] = 0x00; // Empty
+            res = RunUltralightCommand();
+            if (res <= 0)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Try to read a NDEF Message from a Mifare card
+        /// </summary>
+        /// <param name="message">The NDEF message</param>
+        /// <returns>True if success</returns>
+        public bool TryReadNdefMessage(out NdefMessage message)
+        {
+            const int BlockSize = 16;
+            message = new NdefMessage();
+            // Read page 4
+            BlockNumber = 4;
+            Command = UltralightCommand.Read16Bytes;
+            var res = RunUltralightCommand();
+            if (res <= 0)
+            {
+                return false;
+            }
+
+            // Check if it's the old formating and skip it
+            // 01:03:A0:0C
+            int slice = 0;
+            if ((Data![0] == 0x01) && (Data[1] == 0x03) && (Data[2] == 0xA0) && (Data[3] == 0x0C))
+            {
+                slice = 4;
+            }
+
+            var (start, size) = NdefMessage.GetStartSizeNdef(Data.AsSpan(slice));
+
+            if ((start < 0) || (size < 0))
+            {
+                return false;
+            }
+
+            // calculate the size to read
+            int blocksToRead = (size + start) / BlockSize + ((size + start) % BlockSize != 0 ? 1 : 0);
+
+            // We would have to read more available data than the capacity
+            if ((NdefCapacity + 12) <= blocksToRead * BlockSize)
+            {
+                return false;
+            }
+
+            Span<byte> card = new byte[(blocksToRead + slice / 4) * BlockSize];
+            Data.AsSpan(slice).CopyTo(card);
+
+            byte idxCard = 1;
+            // Decrease by 1 block if we skip first page
+            int block = 8;
+            while (idxCard < (blocksToRead + slice / 4))
+            {
+                BlockNumber = (byte)block; // Safe cast as never more than 255
+                Command = UltralightCommand.Read16Bytes;
+                res = RunUltralightCommand();
+                if (res <= 0)
+                {
+                    return false;
+                }
+
+                Data.CopyTo(card.Slice(idxCard * BlockSize - slice));
+                idxCard++;
+                // Here, we read 4 blocks of 4, so 16 blocks
+                block += 4;
+            }
+
+            var ndef = NdefMessage.ExtractMessage(card);
+
+            try
+            {
+                message = new NdefMessage(ndef);
+            }
+            catch (Exception)
```

Comment:
consider inheriting all expected exception from the common base class and catching only that type here - errors like NullRefernceException or ArgumentException should show here still because they mean library error

Reply:
There are really quite a lot which can happen including some null references as explained in the comment. So this is really for convenience and to make sure nothing is not catch. I know it's not fully aligned with the good pattern but I feel it's going to be more confusing if I have a list of 6 or even more possible exceptions.